### ISSUE TITLE
Crash when adding converters to 'list' type (splines)

### ISSUE THREAD

**Issue by Svarov-RZM** [synfig/synfig#1731]
**Issue description**
In some specific cases Synfig can no longer process converters on splines parameters. For example, we have two shapes: rectangle and circle. We want a dot to follow 'rectangle' spline and then half-through the animation jump to 'circle' shape. It's easy to achieve using 'Spline Vertex' converter, we convert origin of a dot and then convert spline parameter to a switch and that's it. 
And this scheme works in Synfig 1.0.2 but doesn't in all newer versions starting from 1.2.x.

**To reproduce**

1. Create any layer with 'Origin' parameter, let's say Circle.
2. Convert 'Origin' to 'Spline Vertex'.
3. Convert 'Spline' parameter of 'Spline Vertex' converter to 'Switch' and Synfig crashes.

It seems to be a global issue and probably affect other converters that can work with splines such as 'Time Loop' for instance but I haven't checked that. Additionally, I attached an example file of the animation described above (it crashes on loading).
[VertexListBug.zip](https://github.com/synfig/synfig/files/5317738/VertexListBug.zip)

**Comment by rodolforg** [comment]
Confirmed :(

**Comment by Svarov-RZM** [comment]
I started learning C++ not so long ago and was trying to solve this issue because it hinders my workflow. I was able to solve it, but not entirely sure I did it correctly.

Please check out this commit: https://github.com/Svarov-RZM/synfig/commit/baae32c80ef91960306cff76ac746ec4fc498afc
I don't know why, but for some reason ValueNode_BLineCalcVertex::operator()(Time t)const in valuenode_blinecalcvertex.cpp was trying to access its own class through pointer (`->` operator) instead of accessing it directly (`.` operator). In short, there were some unnecessary pointer magic and in some cases it would crash in `assert` call.

I reworked it to use direct access to ValueNode_BLineCalcVertex class and program doesn't crash anymore. I also did some testing and didn't find any regressions.

@rodolforg @ice0 what's your opinion on this? Is it appropriate fix?

**Comment by ice0** [comment]
Hi! Thanks for the investigating the problem.

 I haven't dug deep on ValueNodes, so this is an assumption. But as I see, when you convert the Spline parameter to Switch, it cannot convert it to the `ValueNode_BLine` class. The reason it wants to cast is just to check if the spline is looped (Bline is a spline in this context). And it fails because the conversion failed. In this situation, we can assume that the shape not looped.

So here you need to check if the looped parameter is change animation behavior or not. If not, here's how I suggest to change the code:

```
	handle<ValueNode_BLine> bline_value_node( handle<ValueNode_BLine>::cast_dynamic(bline_) );
	//assert(bline_value_node);

	// For classes that cannot be converted to ValueNode_BLine, we assume they are not looped
	const bool looped = bline_value_node ? bline_value_node->get_loop() : false;
```

**Comment by Svarov-RZM** [comment]
> The reason it wants to cast is just to check if the spline is looped

As I understand, there's no need for that, because `loop` parameter is available from ValueBase class, which we can access directly (just like we can access `get_list` from dereferencing *bline_ pointer).


> So here you need to check if the looped parameter is change animation behavior or not

Yes, it does change the animation as expected. I compared against 1.0.2 and 1.2.2 versions.
I am going to run some more tests on weekends and if nothing pop up, I'll send a pull request.

**Comment by rodolforg** [comment]
I checked the [log history of this file](https://github.com/synfig/synfig/commits/master/synfig-core/src/synfig/valuenodes/valuenode_blinecalcvertex.cpp):

There are a couple of changes after v.1.0.2:

- ceaaafad558e27f70e646ea5a06c4613a7c223a2 (pre 1.2.0)
- 5eb54b82e3d7e9860089bfb4635c7ffa52784d09 (pre 1.3.0)
- a90d014c56f0cbee21c2ab0ef68936b9e84c49b1 (pre 1.3.12)

I'll try to `git bisect` it, if the first one is not the culprit.

@Svarov-RZM  thank you for investigate it! If you're correct in the solution, other two files valuenode_blinecalc* should be correct too.

**Comment by rodolforg** [comment]
Argh, how horrible was to build Synfig in those old days.
 I give up lol

Anyway, are you able to switch the spline for the circle origin without crashing?

**Comment by ice0** [comment]
> > The reason it wants to cast is just to check if the spline is looped
> 
> As I understand, there's no need for that, because `loop` parameter is available from ValueBase class, which we can access directly (just like we can access `get_list` from dereferencing *bline_ pointer).

To get `ValueBase` from `ValueNode`, you pass time to it. Thus, the possible `loop` value can change over time (and may be this will be more correct behavior). On the other hand, any solution would be better than a crash )

**Comment by Svarov-RZM** [comment]
>  If you're correct in the solution, other two files valuenode_blinecalc* should be correct too

You're right! The same crash happens in 'Spline Tangent' and 'Spline Width'. The 'Spline Width' converter is broken even more, it doesn't allow you to make conversion, throws an error on attempt such as this:
![SplineWidthBug](https://user-images.githubusercontent.com/28569334/97078968-f451b400-15f8-11eb-8671-244085131064.PNG)
Although, it does allow conversion in 1.2.2 version, so regression was introduced recently.

> Anyway, are you able to switch the spline for the circle origin without crashing?

Yes, of course, I stated above that my commit solves the issue.

> To get `ValueBase` from `ValueNode`, you pass time to it. Thus, the possible `loop` value can change over time (and may be this will be more correct behavior).

Could be probably done in the future, but currently GUI doesn't allow you to animate internal 'loop' parameter. For example, if you enter 'animation' mode, change time to 1s and then click 'Unloop' on any slpine it won't create a waypoint, but simply change parameter for the whole animation.

I sent a pull request. I propose to merge it, wait for regression tests to complete (synfig-tests-regressions) and if everything is alright then we can apply this fix for 'Spline Tangent' and 'Spline Width' converters.

**Comment by rodolforg** [comment]
This issue is already fixed in v1.4.1.

**Comment by morevnaproject** [comment]
The fix is shipped in Synfig Studio version 1.4.2 - https://www.synfig.org/2021/07/29/synfig-studio-1-4-2-released/
